### PR TITLE
Worker daemons set working directory only during execution

### PR DESCRIPTION
**Pull request by ghale**
This is preparatory work for pulling `WorkerDaemonFactory` up to gradle user home scope.  This allows worker daemons to use a project-agnostic "idle" working directory and only switch to the configured working directory while work execution is in progress (switching back to the "idle" working directory when a work item completes).

### CODE REVIEW

File: subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java [hunk 1]
```diff
@@ -73,15 +79,27 @@ public String getDisplayName() {
...
+
+        public File getIdleWorkingDir() {
+            return idleWorkingDir;
+        }
     }
 
     public static class CompilerDaemonServer implements WorkerProtocol<WorkerCompileSpec<?>> {
         @Override
         public DefaultWorkResult execute(WorkerCompileSpec<?> spec) {
+            ProcessEnvironment processEnvironment = NativeServices.getInstance().get(ProcessEnvironment.class);
```
Comment (oehme) [inline subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java]: ⭕️ I wonder if this and `WorkerServer` could be unified?

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/toolchain/PlayToolChainServiceRegistry.java [hunk 2]
```diff
@@ -38,8 +39,8 @@ public void registerProjectServices(ServiceRegistration registration) {
     }
 
     private static class ProjectScopeCompileServices {
-        PlayToolChainInternal createPlayToolChain(FileResolver fileResolver, WorkerDaemonFactory workerDaemonFactory, ConfigurationContainer configurationContainer, DependencyHandler dependencyHandler, WorkerProcessFactory workerProcessBuilderFactory) {
-            return new DefaultPlayToolChain(fileResolver, workerDaemonFactory, configurationContainer, dependencyHandler, workerProcessBuilderFactory);
+        PlayToolChainInternal createPlayToolChain(FileResolver fileResolver, WorkerDaemonFactory workerDaemonFactory, WorkerDirectoryProvider workerDirectoryProvider, ConfigurationContainer configurationContainer, DependencyHandler dependencyHandler, WorkerProcessFactory workerProcessBuilderFactory) {
+            return new DefaultPlayToolChain(fileResolver, workerDaemonFactory, workerDirectoryProvider, configurationContainer, dependencyHandler, workerProcessBuilderFactory);
```
Comment (oehme) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/toolchain/PlayToolChainServiceRegistry.java]: ⭕️ It feels odd that we need to pass both a `WorkerProcessBuilderFactory`(what a name)  and a `WorkerDirectoryProvider`. Shouldn't the former know about the latter and use it internally?

File: subprojects/workers/src/main/java/org/gradle/workers/internal/ActionExecutionSpec.java [hunk 3]
```diff
@@ -32,15 +32,15 @@
 public class ActionExecutionSpec implements WorkSpec {
     private final String displayName;
     private final Class<? extends Runnable> implementationClass;
-    private final File defaultDir;
-    private final File workingDir;
+    private final File idleWorkingDir;
```
Comment (oehme) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/ActionExecutionSpec.java]: ❌ Why would we send this with every request? Shouldn't this be known at worker startup?

Comment (ghale) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/ActionExecutionSpec.java]: Well, I agree this isn't the best.

The problem is that `WorkerDaemonServer` is instantiated by the `WorkerProcess` infrastructure which relies on the server implementations to have no argument constructors.  Just sending this value with the spec seemed to be the simplest, least-worst solution.  

I've changed this to use a dependency injecting instantiator instead and inject an instance of `WorkerDirectoryProvider` on the worker process side (which is probably what I should have done here in the first place).

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/toolchain/PlayToolChainServiceRegistry.java (hunk 2 above)
Comment (ghale) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/toolchain/PlayToolChainServiceRegistry.java]: Yeah, by changing the idle directory to be injected on the server side, this extra service is no longer needed (which cleans this up a good bit).

File: subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java (hunk 1 above)
Comment (ghale) [inline subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java]: Yes, see https://github.com/gradle/gradle/compare/pm/workers/compilation-workers-spike#diff-0ba58c695ab35c11d651adb85c855323.  I consider this to be part of the work to dog-food the worker api in compile tasks.

Comment (ghale) [inline subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java]: Actually, I just went ahead and did this since it cleaned some other things up.

**Comment by ghale** [comment]
Performance tests look good.

**Comment by oehme** [comment]
I'm not sure I understand the last change. How come the compiler daemons don't have to serialize their parameters?

**Comment by ghale** [comment]
Because the parameters in the spec are internal to Gradle, so the classloader in the messaging infrastructure already has what it needs to deserialize it (i.e. it doesn't have to reach through the filtered classloader to get the parameter classes).  This is how the `WorkerCompileSpec` was working before. 

With the worker API, the parameter classes could come from somewhere else completely, and can reference classes in other packages, etc.  If we tried to deserialize the parameters in the classloader in the messaging infrastructure, we'd have to poke enough holes in the filtered classloader to be able to deserialize the parameters _as well as any other classes the parameters touch_.  Instead we just wait until we are in the application classloader (which has unfiltered access to the classpath in the worker configuration) and then deserialize the parameters when we have everything that we need.

In other words, not serializing the compiler parameters in the spec is an optimization we can do because we know those parameters are internal to Gradle.  I hope that makes sense.

**Comment by oehme** [comment]
So if I understand correctly, a next step would be to unify this behavior and always deserialize the action in the application classloader instead of trying to do so earlier. Then we could handle both the same way again. Does that sound reasonable?

**Comment by ghale** [comment]
Well, we could unify this now, but we'd take a small hit to performance - there appeared to be about a 40ms hit on the "large" Java scenarios.  The only reason I made this change was to remove that overhead.  Long term, we should probably take a good look at how this classloader hierarchy works and see if there is anything we could do to handle this problem more efficiently.

**Comment by oehme** [comment]
> Long term, we should probably take a good look at how this classloader hierarchy works and see if there is anything we could do to handle this problem more efficiently.

Yes, that's what I meant. We'd always use the "Simple" version and deserialize that whole object in the right classloader.
